runtimes/starlarkrt/runtime/builtins.go

```diff
@@ -5,107 +5,181 @@ import (
+		"globals":  globalsBuiltinFunc,
+		"module":   starlarkstruct.MakeModule,
+		"struct":   starlarkstruct.Make,
+	}
+)
+
+func activityBuiltinFunc(th *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	if len(args) == 0 {
+		return nil, fmt.Errorf("missing function argument")
+	}
+
+	tlsContext, ok := th.Local(tlsKey).(*tlsContext)
+	if !ok {
+		return nil, fmt.Errorf("context is not set")
+	}
+
+	akV, err := tlsContext.vctx.FromStarlarkValue(args[0])
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert function from starlark: %w", err)
+	}
+
+	akArgs, err := kittehs.TransformError(args[1:], tlsContext.vctx.FromStarlarkValue)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert args from starlark: %w", err)
+	}
+
+	akKwArgs, err := kittehs.ListToMapError(kwargs, func(t starlark.Tuple) (string, sdktypes.Value, error) {
+		if len(t) != 2 {
+			return "", sdktypes.InvalidValue, fmt.Errorf("expected 2 elements, got %d", len(t))
+		}
 
-		fn, ok := args[0].(starlark.Callable)
+		k, ok := t[0].(starlark.String)
 		if !ok {
-			return nil, errors.New("first argument must be a function")
+			return "", sdktypes.InvalidValue, fmt.Errorf("expected string, got %T", t[0])
 		}
 
-		value, err := starlark.Call(thread, fn, args[1:], kwargs)
+		v, err := tlsContext.vctx.FromStarlarkValue(t[1])
 		if err != nil {
-			if perr, ok := sdktypes.FromError(err); ok {
-				v := perr.Value()
-				vctx := values.Context{}
-				if slv, cerr := vctx.ToStarlarkValue(v); cerr == nil {
-					return starlark.Tuple{starlark.None, slv}, nil
-				}
-			}
-
-			return starlark.Tuple{starlark.None, starlark.String(err.Error())}, nil
+			return "", sdktypes.InvalidValue, fmt.Errorf("cannot convert value from starlark: %w", err)
 		}
 
-		return starlark.Tuple{value, starlark.None}, nil
-	},
-	"fail": func(thread *starlark.Thread, bi *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
-		vctx := values.Context{}
+		return k.GoString(), v, nil
+	})
+	if err != nil {
+		return nil, err
+	}
 
-		if len(args) != 0 && len(kwargs) != 0 {
-			return nil, fmt.Errorf("cannot specify both positional and keyword arguments")
-		}
+	rv, err := tlsContext.cbs.Call(tlsContext.goCtx, tlsContext.runID, akV, akArgs, akKwArgs)
+	if err != nil {
+		return nil, err
+	}
 
-		if len(args) == 1 {
-			v, err := vctx.FromStarlarkValue(args[0])
-			if err != nil {
-				return nil, fmt.Errorf("cannot convert value from starlark: %w", err)
-			}
-			return nil, sdktypes.NewProgramError(v, nil, nil).ToError()
-		}
+	return tlsContext.vctx.ToStarlarkValue(rv)
+}
 
-		if len(args) > 1 {
-			vs, err := kittehs.TransformError(args, vctx.FromStarlarkValue)
-			if err != nil {
-				return nil, fmt.Errorf("cannot convert values from starlark: %w", err)
-			}
+func catchBuiltinFunc(th *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
```

Comment:
How is this related to the PR?

Reply:
did a little refactoring that simplified the structure as part of it.